query/src/sql/planner/plan/mod.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod logical_get;
+mod logical_project;
+mod physical_project;
+mod physical_scan;
+
+pub use logical_get::LogicalGet;
+pub use logical_project::LogicalProject;
+pub use logical_project::ProjectItem;
+pub use physical_project::PhysicalProject;
+pub use physical_scan::PhysicalScan;
+
+use crate::sql::optimizer::PhysicalProperty;
+use crate::sql::optimizer::RelationalProperty;
+use crate::sql::optimizer::RequiredProperty;
+use crate::sql::optimizer::SExpr;
+
+pub trait LogicalPlan {
+    fn compute_relational_prop(&self, expression: &SExpr) -> RelationalProperty;
+
+    fn as_plan(&self) -> Plan;
+}
+
+pub trait PhysicalPlan {
+    fn compute_physical_prop(&self, expression: &SExpr) -> PhysicalProperty;
+
+    fn compute_required_prop(&self, input_prop: &RequiredProperty) -> RequiredProperty;
+
+    fn as_plan(&self) -> Plan;
+}
+
+/// Relational operator
+#[derive(Clone, PartialEq, Debug)]
+pub enum Plan {
+    // Logical operators
+    LogicalGet(LogicalGet),
+    LogicalProject(LogicalProject),
+
+    // Physical operators
+    PhysicalScan(PhysicalScan),
+    PhysicalProject(PhysicalProject),
```

Comment:
Will it make Plan too fatter if we have logical/physical plan in a single enum?

Visitor/Rewriter should consider about all match cases.

Reply:
We have requirement for down casting a base `Plan` type to concrete logical or physical plan struct.

For example:

https://github.com/datafuselabs/databend/blob/0f17720ddfe8797a87523063c5a5629697b81d44/query/src/sql/optimizer/rule/rule_implement_get.rs#L46-L53

I found that enum is the best choice in rust to implement this.

The good thing is we don't need to write too many visitors to probe the `Plan` with this framework. Every `Rule` can focus on the subtree matches its pattern expression.